array init error

**V version:** 
0.1.27

**OS:** 
Linux

Something wrong with the 'init' identifier when building the array; not having it will solve the compilation error.


```
module main

struct MyStruct {
pub mut:
	ar []f32
}

fn main() {
	mut m := MyStruct {
		ar: []f32{len: 10, init:0}
	}
	println(m.ar[0])
}
```

```
$ v run array_bug.v 
==================
/home/roso/.cache/v/array_bug.tmp.c: In function ‘main’:
/home/roso/.cache/v/array_bug.tmp.c:9734:54: error: ‘_val_92’ undeclared (first use in this function)
 9734 |  .ar = __new_array_with_default(10, 0, sizeof(f32), &_val_92),
      |                                                      ^~~~~~~
/home/roso/.cache/v/array_bug.tmp.c:9734:54: note: each undeclared identifier is reported only once for each function it appears in
/home/roso/.cache/v/array_bug.tmp.c: In function ‘vcalloc’:
/home/roso/.cache/v/array_bug.tmp.c:4728:1: warning: control reaches end of non-void function [-Wreturn-type]
 4728 | }
      | ^
/home/roso/.cache/v/array_bug.tmp.c: In function ‘byte_is_white’:
/home/roso/.cache/v/array_bug.tmp.c:7499:1: warning: control reaches end of non-void function [-Wreturn-type]
 7499 | }
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -967,9 +967,7 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 			if is_call {
 				g.expr(val)
 			} else {
-				if val is ast.ArrayInit {
-					g.gen_default_init_value(val as ast.ArrayInit)
-				}
+				g.gen_default_init_value(val)
 				g.write('{$styp _ = ')
 				g.expr(val)
 				g.writeln(';}')
@@ -978,9 +976,7 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 			right_sym := g.table.get_type_symbol(assign_stmt.right_types[i])
 			mut is_fixed_array_init := false
 			mut has_val := false
-			if val is ast.ArrayInit {
-				is_fixed_array_init, has_val = g.gen_default_init_value(val as ast.ArrayInit)
-			}
+			is_fixed_array_init, has_val = g.gen_default_init_value(val)
 			is_inside_ternary := g.inside_ternary != 0
 			cur_line := if is_inside_ternary {
 				g.register_ternary_name(ident.name)
@@ -1045,16 +1041,28 @@ fn (mut g Gen) gen_assign_stmt(assign_stmt ast.AssignStmt) {
 	}
 }
 
-fn (mut g Gen) gen_default_init_value(val ast.ArrayInit) (bool, bool) {
-	is_fixed_array_init := val.is_fixed
-	has_val := val.has_val
-	elem_type_str := g.typ(val.elem_type)
-	if val.has_default {
-		g.write('$elem_type_str _val_$val.pos.pos = ')
-		g.expr(val.default_expr)
-		g.writeln(';')
-	} else if val.has_len && val.elem_type == table.string_type {
-		g.writeln('$elem_type_str _val_$val.pos.pos = tos_lit("");')
+fn (mut g Gen) gen_default_init_value(val ast.Expr) (bool, bool) {
+	mut is_fixed_array_init := false
+	mut has_val := false
+	match val {
+		ast.ArrayInit {
+			is_fixed_array_init = it.is_fixed
+			has_val = it.has_val
+			elem_type_str := g.typ(it.elem_type)
+			if it.has_default {
+				g.write('$elem_type_str _val_$it.pos.pos = ')
+				g.expr(it.default_expr)
+				g.writeln(';')
+			} else if it.has_len && it.elem_type == table.string_type {
+				g.writeln('$elem_type_str _val_$it.pos.pos = tos_lit("");')
+			}
+		}
+		ast.StructInit {
+			for field in it.fields {
+				g.gen_default_init_value(field.expr)
+			}
+		}
+		else {}
 	}
 	return is_fixed_array_init, has_val
 }

--- a/vlib/v/tests/array_init_test.v
+++ b/vlib/v/tests/array_init_test.v
@@ -127,3 +127,16 @@ fn test_array_string_full_options() {
 	assert c.cap >= c.len
 	assert c.str() == "['a', 'b', 'c', 'c', 'c', 'c', 'c', 'c', 'c', 'c', '11', '12', '13', '14', '15', '16']"
 }
+
+struct MyStruct {
+pub mut:
+	ar []f32
+}
+
+fn test_array_init_in_struct_field() {
+	m := MyStruct {
+		ar: []f32{len: 4, init:1.2}
+	}
+	println(m)
+	assert m.ar.str() == '[1.2, 1.2, 1.2, 1.2]'
+}

